### PR TITLE
Api Documentation: easyscore

### PR DESCRIPTION
**Pull request by rvilarl**
@0xfe @ronyeh I am trying to render Claire de Lune, documenting the Easy Score API on the way. I did not get too far and I suggest to use this PR to discuss and progress on the documentation.

### CODE REVIEW

**Comment by rvilarl** [comment]
```js
const vf: Factory = new Vex.Flow.Factory({renderer: { elementId: 'boo', width: 1200, height: 8000 }});
const score: EasyScore = vf.EasyScore();
score.set({ time: '9/8' });
const system: System = vf.System({ x: x, y: y, width: measure.width });
system.addStave({ voices: [score.voice(score.notes('C4/8/r, C4/8/r, (F5 Ab5)/8, (F5 Ab5)/q, (Db5 F5)/q'), {})] })
            .addClef('treble')
            .addKeySignature('Db')
            .addTimeSignature('9/8');
```
image
@0xfe I have the following doubts:
- What is the right way to center the rests? Best not to provide location(ie C4) they are normally not provided on MusicXML
- How should I get that the note representation attend the key signature?

**Comment by ronyeh** [comment]
If you look at my PR for rests tests, there is an option for auto align.
Can you set that option via easy score?

**Comment by rvilarl** [comment]
> If you look at my PR for rests tests, there is an option for auto align. Can you set that option via easy score?

it seems that `System` does not accept `FormatOptions` and there is `align_rest`

**Comment by rvilarl** [comment]
@0xfe @ronyeh my proposal is to  resolve the two questions above, merge and continue with the next problem that I get rendering the partiture. Next step will be second staff of the first measure.

**Comment by rvilarl** [comment]
> > If you look at my PR for rests tests, there is an option for auto align. Can you set that option via easy score?
> 
> it seems that `System` does not accept `FormatOptions` and there is `align_rest`

@ronyeh thans for the hint!
rest alignment issue is now resolved adding FormatOptions to SystemOptions and fixing the grammar in EasyScore to allow R as note name
@0xfe please look into the changes.

**Comment by rvilarl** [comment]
```js
const vf: Factory = new Vex.Flow.Factory({renderer: { elementId: 'boo', width: 1200, height: 8000 }});
const score: EasyScore = vf.EasyScore();
score.set({ time: '9/8' });
const system: System = vf.System({ x: x, y: y, width: measure.width, formatOptions: { align_rests: true } });
system.addStave({ voices: [score.voice(score.notes('R4/8/r, R4/8/r, (F5 Ab5)/8, (F5 Ab5)/q, (Db5 F5)/q'), {})] })
            .addClef('treble')
            .addKeySignature('Db')
            .addTimeSignature('9/8');
```
image

@0xfe @ronyeh Now only one question is open:
- What should I do so that the note representation takes into account the key signature? The notes should not render a `b`.

File: src/factory.ts [hunk 1]
```diff
@@ -642,6 +656,17 @@ export class Factory {
...
 
+  /**
+   * Creates EasyScore. Normally the first step after constructing a factory.
+   *
+   * Example:
+   * `const vf: Factory = new Vex.Flow.Factory({renderer: { elementId: 'boo', width: 1200, height: 600 }});`
+   * `const score: EasyScore = vf.EasyScore();`
+   * @param params.factory not required
+   * @param params.builder TODO @0xfe
+   * @param params.commitHooks TODO @0xfe
```
Comment (rvilarl) [inline src/factory.ts:667]: @0xfe what is the aim of these parameters?

File: src/easyscore.ts [hunk 2]
```diff
@@ -162,7 +161,7 @@ export class Grammar {
     return { token: "['][^']*[']" };
   }
   NOTENAME(): Rule {
-    return { token: '[a-gA-G]' };
+    return { token: '[a-gA-GrR]' };
```
Comment (ronyeh) [inline src/easyscore.ts:164]: Does note name = R result in a auto aligned rest? Good idea! EasyScore should definitely be as easy as possible.

Comment (rvilarl) [reply src/easyscore.ts:164]: There is in the code a strange condition to align rests:

https://github.com/0xfe/vexflow/blob/3eea242fc47c7d540b226e969b6ad6a984ee5f7e/src/formatter.ts#L347

The position must be B/4 or R/4. I prefered to put support to R in easyscore. It is also quite strange that they mus be in the octave 4. 

In order to get it working, you also have to formatOptions like in the example below:

```js
const system: System = vf.System({ x: x, y: y, width: measure.width, formatOptions: { align_rests: true } });
```

your idea is very good, R => auto align rest. @0xfe what do you think?

Comment (0xfe) [reply src/easyscore.ts:164]: I'm okay with the distinction, but I'm not sure overloading `NOTENAME` is the right approach. There's `TYPE | TYPES` below which does rests.

Comment (0xfe) [reply src/easyscore.ts:164]: Also, can we separate this change from the documentation?

**Comment by rvilarl** [comment]
@0xfe how do I get the `b` in the notes not displayed and attending to the time signature?

File: src/easyscore.ts (hunk 2 above)
Comment (rvilarl) [reply src/easyscore.ts]: Done

File: src/stave.ts [hunk 3]
```diff
@@ -504,14 +505,35 @@ export class Stave extends Element {
     return this;
   }
 
-  addKeySignature(keySpec: string, cancelKeySpec: string, position?: number): this {
+  /**
+   * Add a key signature to the stave.
+   *
+   * Example:
+   * `stave.addKeySignature('Db');`
+   * @param keySpec new key specitication `[A-G][b|#]?`
```
Comment (ronyeh) [inline src/stave.ts:513]: typo: specitication

File: src/factory.ts (hunk 1 above)
Comment (0xfe) [reply src/factory.ts:667]: `EasyScore` uses a `Parser` to parse EasyScore `Grammar` and a `Builder` to actually build the VF elements.

A `commitHook` is like an extension, that's called to set additional properties on an element (usually a note) after it is built. E.g., `setId` sets the `id` of a note.

Comment (0xfe) [reply src/factory.ts:667]: The builder uses the `Factory` to build the elements, it doesn't build them directly. (This also keeps the Factory API consistent.)

**Comment by rvilarl** [comment]
> _(note)_
> _Looking at the testing I tried:_
> 
> ```js
> const vf: Factory = new Vex.Flow.Factory({renderer: { elementId: 'boo', width: 1200, height: 8000 }});
> const score: EasyScore = vf.EasyScore();
> score.set({ time: '9/8' });
> const system: System = vf.System({ x: x, y: y, width: measure.width, formatOptions: { align_rests: true } });
> const voice = score.voice(score.notes('R4/8/r, R4/8/r, (F5 Ab5)/8, (F5 Ab5)/q, (Db5 F5)/q'), {}).setStrict(false);
> Vex.Flow.Accidental.applyAccidentals([voice], 'Db');
> system.addStave({ voices: [voice] })
>             .addClef('treble')
>             .addKeySignature('Db')
>             .addTimeSignature('9/8');
> ```
> 
> _But obviously it is not the solution :(_
> 
> image

@0xfe @ronyeh I believe that this should be indeed the solution but that there is a bug in `EasyScore.notes`. The StaveNotes returned by this function do not include the accidentals in the attribute `keys` and this is why `applyAccidentals` does not work.

With the following code:

```js
var notes = [
        { keys: ['a/4'], duration: '4' },
        { keys: ['b/4'], duration: '4' },
        { keys: ['c#/5'], duration: '4' },
        { keys: ['d/5'], duration: '4' },
        { keys: ['e/5'], duration: '4' },
        { keys: ['f#/5'], duration: '4' },
        { keys: ['g#/5'], duration: '4' },
        { keys: ['a/5'], duration: '4' },
      ].map(vf.StaveNote.bind(vf));

      const score = vf.EasyScore();
      var notes1 = score.notes('A4/q, B4/q, C#5/q, D5/q, E5/q, F#/5, G#/5, A5/q');
```
**notes[2].keys is c#/5 however notes1[2].keys is C/5**

**Comment by rvilarl** [comment]
@0xfe @ronyeh problem solved in last commit, `applyAccidentals` works. 
(the code below is generated programatically from a musicxml score)

```js
const vf: Factory = new Vex.Flow.Factory({renderer: { elementId: myCanvas, width: 1190.88, height: 3366.72 }});
const score: EasyScore = vf.EasyScore();
score.set({ time: "9/8" });
system = vf.System({x: 77.69290000000001, y: 170, width: 356.68, formatOptions: { align_rests: true }});
notes = notes.concat(score.notes(B4/8/r, B4/8/r, (F5 Ab5)/8, (F5 Ab5)/q., (Db5 F5)/q.));
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
stave.addClef("treble").addKeySignature("Db").addTimeSignature("9/8");
notes = notes.concat(score.notes(B4/8/r, (F4 Ab4)/q, (F4 Ab4)/h.));
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
stave.addClef("treble").addKeySignature("Db").addTimeSignature("9/8");
system.addConnector("brace");
system.addConnector("singleLeft");
system.addConnector("singleRight");
vf.draw();
system = vf.System({x: 434.3729, y: 170, width: 221.48, formatOptions: { align_rests: true }});
notes = notes.concat(score.notes(""));
notes = notes.concat(score.beam(score.notes("(Db5 F5)/8, (C5 Eb5)/8, (Db5 F5)/8"), { autoStem: true }));
notes = notes.concat(score.notes((C5 Eb5)/h.));
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
notes = notes.concat(score.notes((Gb4 A4)/h., (Gb4 A4)/q.));
voices.push(score.voice(notes, {}).setStrict(false));
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
system.addConnector("singleRight");
vf.draw();
system = vf.System({x: 655.8529, y: 170, width: 275.38, formatOptions: { align_rests: true }});
notes = notes.concat(score.notes(""));
notes = notes.concat(score.beam(score.notes("(C5 Eb5)/8, (Bb4 Db5)/8, (C5 Eb5)/8"), { autoStem: true }));
notes = notes.concat(score.notes("(Bb4 Db5)/8., F5/q"));
notes = notes.concat(score.beam(score.notes("F5/8, Db5/8."), { autoStem: true }));
notes = notes.concat(score.notes());
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
notes = notes.concat(score.notes((F4 Ab4)/h., (F4 Ab4)/q.));
voices.push(score.voice(notes, {}).setStrict(false));
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
system.addConnector("singleRight");
vf.draw();
system = vf.System({x: 931.2329, y: 170, width: 202.95, formatOptions: { align_rests: true }});
notes = notes.concat(score.notes(""));
notes = notes.concat(score.beam(score.notes("Db5/8, (Ab4 C5)/8, (Bb4 Db5)/8"), { autoStem: true }));
notes = notes.concat(score.notes((Ab4 C5)/h.));
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
notes = notes.concat(score.notes((Eb4 Gb4)/h., (Eb4 Gb4)/q.));
voices.push(score.voice(notes, {}).setStrict(false));
notes = [];
Vex.Flow.Accidental.applyAccidentals(voices, 'Db');
const stave = system.addStave({ voices: voices });
voices = [];
system.addConnector("singleRight");
vf.draw();
```

image

@AaronDavidNewman I believe that the notes on the upper stave in the 2nd and 4th compasses are too close, wouldn't you agree?

File: src/stave.ts (hunk 3 above)
Comment (rvilarl) [reply src/stave.ts]: done

File: src/factory.ts (hunk 1 above)
Comment (rvilarl) [reply src/factory.ts:667]: Thanks, what would you write in the comment then?

**Comment by rvilarl** [comment]
I would like to go for a merge now and continue with Debussy until all is presented correctly.

**Comment by AaronDavidNewman** [comment]
It does look squished.  You can open a different PR against the debussy issue.  It looks like the issue is with padding - modifiers add the padding they need, but there are no modifiers, so maybe it has to do with that.

File: src/easyscore.ts [hunk 4]
```diff
@@ -367,11 +383,12 @@ export class Builder {
     chord.forEach((notePiece: NotePiece, index: number) => {
       const accid = notePiece.accid;
       if (typeof accid === 'string') {
-        const accidental: Accidental = factory.Accidental({ type: accid });
-        // TODO: Remove "as unknown as Modifier".
-        // This compilation warning will be fixed after factory & accidental are migrated to typescript.
-        note.addAccidental(index, accidental as unknown as Modifier);
-        accidentals.push(accidental);
+        // Do not attach sharps or flats they are included in keys.
+        if (sharpFlat(accid) == '') {
```
Comment (0xfe) [inline src/easyscore.ts:387]: I don't understand what's going on here. Can you add some tests?

Also, can you put this bugfix in a different PR? Let's keep this PR for just the documentation updates.

File: src/factory.ts (hunk 1 above)
Comment (0xfe) [reply src/factory.ts:667]: No need to be too detailed here. This is sufficient:

builder: Instance of Builder
commitHooks: Function to call after a note element is created

File: src/easyscore.ts (hunk 4 above)
Comment (0xfe) [reply src/easyscore.ts:387]: I think this will break EasyScore. Did you check the regression tests?

Comment (rvilarl) [reply src/easyscore.ts:387]: You are right,  I was yesterday so focused on getting the applyAccidentals right that I broke the other ones. I assumed that I was right but there is something missing.

File: src/factory.ts (hunk 1 above)
Comment (rvilarl) [reply src/factory.ts]: done

**Comment by rvilarl** [comment]
> It does look squished. You can open a different issue against the that. It might be the padding - modifiers add the padding they need, but there are no modifiers, so maybe it has to do with that.

Thanks @AaronDavidNewman it might have to do with the `applyAccidentals` correction. I will make the issue if it stays once this is corrected.

File: src/easyscore.ts (hunk 4 above)
Comment (rvilarl) [reply src/easyscore.ts]: removed, I will open an issue to resolve this topic.

**Comment by rvilarl** [comment]
@0xfe Ready to merge. Regression tests pass. :)

**Comment by 0xfe** [comment]
Thanks Rodrigo -- can you fix the conflicts please?

**Comment by rvilarl** [comment]
Done

**Comment by 0xfe** [comment]
Looks good!
